nx/lib/nx.ex

```diff
@@ -323,6 +323,79 @@ defmodule Nx do
   For a more complex slicing rules, including strides, you
   can always fallback to `Nx.slice/4`.
 
+  Access also supports updating values with put_in/3 and update_in/3,
+  where the path to the value is a list of integers indicating the sub-dimensions.
+
+  The value that will be used to substitute should be a tensor of the same
+  shape as the one being updated.
+
+      iex> t = Nx.tensor([1, 2, 3])
+      iex> put_in(t, [2], Nx.tensor(2))
+      #Nx.Tensor<
+        s64[3]
+        [1, 2, 2]
+      >
+
+  In a multi-dimensional array, the path will determine at which level
+  the update will be made.
+
+  A whole tensor can be updated:
+
+      iex> t = Nx.tensor([[1, 2], [3, 4]])
+      iex> put_in(t, [1], Nx.tensor([4, 5]))
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [1, 2],
+          [4, 5]
+        ]
+      >
+
+  Or a scalar in a deeper level:
+
+      iex> t = Nx.tensor([[[1, 2], [3, 4]], [[5, 6], [7, 8]]])
+      iex> put_in(t, [0, 1, 1], Nx.tensor(99))
+      #Nx.Tensor<
+        s64[2][2][2]
+        [
+          [
+            [1, 2],
+            [3, 99]
+          ],
+          [
+            [5, 6],
+            [7, 8]
+          ]
+        ]
+      >
+
+  As long as the update value is the same shape as the target one
+  to be substituted it will replace it, if not it will raise:
+
+      iex> t = Nx.tensor([[1, 2, 3], [4, 5, 6]])
+      iex> put_in(t, [0], Nx.tensor([1, 2]))
+      ** (ArgumentError) the shape {2} of the new value does not match the current value shape {3}
+
+  Out of bound attempts to update will also raise:
+      iex> t = Nx.tensor([[1, 2], [3, 4]])
+      iex> put_in(t, [2], Nx.tensor([5, 6]))
+      ** (ArgumentError) index 2 is out of bounds for axis 0 in shape {2, 2}
+
+  Unlike the access get syntax, it does not accept negative indices.
```

Comment:
On this I am not sure. It seems a bit weird for me to support negative indices to update, but to be consistent with the access syntax used to get values it should be supported. I'll take a look to add it.

Reply:
I would accept negative values as well. I'm not totally sure on the index format, though. I would expect the index path to be more similar to Nx.gather/indexed_add/indexed_put

Reply:
For supporting negative indices, you can do `dimension_size + negative_index` to turn it into the corresponding non-negative index

Reply:
I will add the support for negative indexes then. As for the index format, `Kernel.put_in` does expect a list, I actually tried it so it could be used with both tensors and lists and it doesn't work. I also think this is intuitive enough as the list of keys is what is used with `put_in` and `update_in`.

Reply:
So the idea is to support a list of 1 to n entries, where n is the rank of the tensor being Access-ed, right?

Reply:
Yes. That's a very clear way to put it, I will add it to the docs.

Reply:
Well, n is <= rank, you can also access `[0]` and substitute an N-dimensional array instead of a single entry.

Reply:
What would one expect when updating a nested list of tensors? I'm trying to wrap my head around how things would work

For instance: `[1, 2, [Nx.tensor([[3, 4], [5, 6]]), 7]]`

Valid keys would be like below? I'm representing `key -> value accessed`

```elixir
[0] -> 1
[1] -> 2
[2] -> [Nx.tensor(...), 7]
[2, 0] -> Nx.tensor([[3, 4], [5, 6]])
[2, 0, [0]] -> Nx.tensor([3, 4])
[2, 0, [1]] -> Nx.tensor([5, 6])
[2, 0, [0, 0]] -> Nx.tensor(3)
[2, 0, [1, 1]] -> Nx.tensor(6)
[2, 1] -> 7
```

Reply:
Hmm, well, lists do not implement the Access behaviour, but if it all were tensors then yes it would be similar to that, for example:

```elixir
Nx.tensor([[[1, 2]],  [[3, 4]]])

[0] -> [[1, 2]]
[0, 0] -> [1, 2]
[0, 0, 1] -> 2
[1, 0, 1] -> 4
[0, 1] -> Out of boundaries
```

So for a shape {a, b, c} you can make a list of up to [x, y, z] where x <= a, etc. but you can also do [x, y] and that will give you a tensor of shape {b, c}, meaning for each element of the list to access the rank of the tensor to update will diminish by 1.

Reply:
I forgot to add the `Access.at` calls:

```elixir
iex(3)> l = [1, 2, [3, [4, 5]]]
[1, 2, [3, [4, 5]]]
iex(4)> get_in(l, [Access.at(0)])
1
iex(5)> get_in(l, [Access.at(2), Access.at(2)])
nil
iex(6)> get_in(l, [Access.at(2), Access.at(1)])
[4, 5]
```

Same idea as what I specified there, though.

Reply:
My main concern is on the index format and how you would access a tensor in a nested list, though

Reply:
So yeah, it would be a combination of `Access.at(i)` and integers. I just tried this and it does work:

```elixir
list = [1, 2, Nx.tensor([1, 2, 3])]
put_in(list, [Access.at(2), 2], Nx.tensor(99))

[
  1,
  2,
  #Nx.Tensor<
    s64[3]
    [1, 2, 99]
  >
]
```

Reply:
Yeah, this is starting to make sense to me now :)